### PR TITLE
qt6: avoid building under /tmp on darwin

### PR DESCRIPTION
**Pull request by wegank**
###### Description of changes

https://github.com/NixOS/nixpkgs/pull/224469#issuecomment-1496299746

🤮

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by NickCao** [comment]
So who is to blame, qt or cmake?

**Comment by wegank** [comment]
I would say CMake, since according to https://github.com/qt/qtbase/commit/41aff1cd938457539fd6fd1009a3859de6b78095:

```
Building Qt in a symlinked build directory leads to all kind of issues,
like unnecessary rebuilds, strange compilation errors or already errors
at configure time.
```

It sufficed to remove the check for building Qt 6.4.2 (expect Qt WebEngine), but it's not the case for Qt 6.5.0. Fortunately, we can build it under `/nix/store`, which is not symlinked. Good luck to the Homebrew guys...

---

I converted this PR to draft, since this hack is also necessary to build Qt Declarative 6.5.0.

```
error: builder for '/nix/store/lb89cxk9aq9waa41cypa6p4pahy4ds91-qtdeclarative-6.5.0.drv' failed with exit code 1;
       last 10 log lines:
       > In file included from /tmp/nix-build-qtdeclarative-6.5.0.drv-0/qtdeclarative-everywhere-src-6.5.0/src/qml/../3rdparty/masm/config.h:43:
       > In file included from /tmp/nix-build-qtdeclarative-6.5.0.drv-0/qtdeclarative-everywhere-src-6.5.0/src/qml/../3rdparty/masm/wtf/Platform.h:33:
       > In file included from /tmp/nix-build-qtdeclarative-6.5.0.drv-0/qtdeclarative-everywhere-src-6.5.0/build/include/QtQml/6.5.0/QtQml/private/qv4global_p.h:1:
       > In file included from /tmp/nix-build-qtdeclarative-6.5.0.drv-0/qtdeclarative-everywhere-src-6.5.0/build/include/QtQml/6.5.0/QtQml/private/../../../../../../src/qml/jsruntime/qv4global_p.h:22:
       > In file included from /tmp/nix-build-qtdeclarative-6.5.0.drv-0/qtdeclarative-everywhere-src-6.5.0/build/include/QtQml/qtqmlglobal.h:1:
       > /tmp/nix-build-qtdeclarative-6.5.0.drv-0/qtdeclarative-everywhere-src-6.5.0/build/include/QtQml/../../../src/qml/qtqmlglobal.h:9:10: fatal error: 'QtQml/qtqml-config.h' file not found
       > #include <QtQml/qtqml-config.h>
       >          ^~~~~~~~~~~~~~~~~~~~~~
       > 1 error generated.
       > ninja: build stopped: subcommand failed.
       For full logs, run 'nix log /nix/store/lb89cxk9aq9waa41cypa6p4pahy4ds91-qtdeclarative-6.5.0.drv'.
```

File: pkgs/build-support/setup-hooks/drop-build-tree.sh [hunk 1]
```diff
@@ -0,0 +1,5 @@
+postPhases+=" removeBuildDir"
+
+removeBuildDir() {
+    rm -rf $out/.build
```
Comment (NickCao) [inline pkgs/build-support/setup-hooks/drop-build-tree.sh:4]: It feels that this should be part of `keepBuildTree`.

Comment (wegank) [reply pkgs/build-support/setup-hooks/drop-build-tree.sh:4]: I guess it should not...? `dropBuildTree` means to cancel the effect of  `keepBuildTree` on packages not depending on `$NIX_BUILD_TOP`.

Comment (NickCao) [reply pkgs/build-support/setup-hooks/drop-build-tree.sh:4]: There is an implicit contract on the path `$out/.build` between the two hooks (and there's no reference to `keepBuildTree` treewide), so IMO they'd better be located in the same `.sh`.

**Review by NickCao — APPROVED** [review]
Diff LGTM, I'd like to integrate this directly into https://github.com/NixOS/nixpkgs/pull/224469 to avoid adding the version constraints, also waiting for https://github.com/NixOS/nixpkgs/pull/226018 as it may fix a wide range of problems we are currently having with qt6 packaging.

**Comment by wegank** [comment]
@ofborg build python311Packages.pyqt6

**Comment by wegank** [comment]
Result of `nixpkgs-review pr 226033` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>24 packages marked as broken and skipped:</summary>
  <ul>
    <li>adwaita-qt6</li>
    <li>adwaita-qt6.dev</li>
    <li>anki</li>
    <li>anki.dist</li>
    <li>anki.doc</li>
    <li>anki.man</li>
    <li>mnemosyne</li>
    <li>mnemosyne.dist</li>
    <li>olive-editor</li>
    <li>python310Packages.pyqt6-webengine</li>
    <li>python310Packages.pyqt6-webengine.dev</li>
    <li>python310Packages.pyqt6-webengine.dist</li>
    <li>python311Packages.pyqt6-webengine</li>
    <li>python311Packages.pyqt6-webengine.dev</li>
    <li>python311Packages.pyqt6-webengine.dist</li>
    <li>qt6.qtwayland</li>
    <li>qt6.qtwayland.dev</li>
    <li>qt6Packages.qtwayland</li>
    <li>qt6Packages.qtwayland.dev</li>
    <li>qutebrowser-qt6</li>
    <li>qutebrowser-qt6.dist</li>
    <li>retext</li>
    <li>retext.dist</li>
    <li>stellarium</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>alice-tools-qt6</li>
  </ul>
</details>
<details>
  <summary>93 packages built:</summary>
  <ul>
    <li>bambootracker-qt6</li>
    <li>beamerpresenter (beamerpresenter-mupdf)</li>
    <li>beamerpresenter-poppler</li>
    <li>cutemaze</li>
    <li>dolphin-emu</li>
    <li>gpxsee-qt6</li>
    <li>hydrus</li>
    <li>hydrus.doc</li>
    <li>khoj</li>
    <li>khoj.dist</li>
    <li>ladybird</li>
    <li>nanovna-saver</li>
    <li>nanovna-saver.dist</li>
    <li>pokefinder</li>
    <li>python310Packages.pyqt6</li>
    <li>python310Packages.pyqt6-charts</li>
    <li>python310Packages.pyqt6-charts.dist</li>
    <li>python310Packages.pyqt6.dev</li>
    <li>python310Packages.pyqt6.dist</li>
    <li>python311Packages.pyqt6</li>
    <li>python311Packages.pyqt6-charts</li>
    <li>python311Packages.pyqt6-charts.dist</li>
    <li>python311Packages.pyqt6.dev</li>
    <li>python311Packages.pyqt6.dist</li>
    <li>qalculate-qt</li>
    <li>qt6.qt3d (qt6Packages.qt3d)</li>
    <li>qt6.qt3d.dev (qt6Packages.qt3d.dev)</li>
    <li>qt6.qt5compat (qt6Packages.qt5compat)</li>
    <li>qt6.qt5compat.dev (qt6Packages.qt5compat.dev)</li>
    <li>qt6.qtbase (qt6Packages.qtbase)</li>
    <li>qt6.qtbase.dev (qt6Packages.qtbase.dev)</li>
    <li>qt6.qtcharts (qt6Packages.qtcharts)</li>
    <li>qt6.qtcharts.dev (qt6Packages.qtcharts.dev)</li>
    <li>qt6.qtconnectivity (qt6Packages.qtconnectivity)</li>
    <li>qt6.qtconnectivity.dev (qt6Packages.qtconnectivity.dev)</li>
    <li>qt6.qtdatavis3d (qt6Packages.qtdatavis3d)</li>
    <li>qt6.qtdatavis3d.dev (qt6Packages.qtdatavis3d.dev)</li>
    <li>qt6.qtdeclarative (qt6Packages.qtdeclarative)</li>
    <li>qt6.qtdeclarative.dev (qt6Packages.qtdeclarative.dev)</li>
    <li>qt6.qtdoc (qt6Packages.qtdoc)</li>
    <li>qt6.qthttpserver (qt6Packages.qthttpserver)</li>
    <li>qt6.qthttpserver.dev (qt6Packages.qthttpserver.dev)</li>
    <li>qt6.qtimageformats (qt6Packages.qtimageformats)</li>
    <li>qt6.qtimageformats.dev (qt6Packages.qtimageformats.dev)</li>
    <li>qt6.qtlanguageserver (qt6Packages.qtlanguageserver)</li>
    <li>qt6.qtlanguageserver.dev (qt6Packages.qtlanguageserver.dev)</li>
    <li>qt6.qtlottie (qt6Packages.qtlottie)</li>
    <li>qt6.qtlottie.dev (qt6Packages.qtlottie.dev)</li>
    <li>qt6.qtmultimedia (qt6Packages.qtmultimedia)</li>
    <li>qt6.qtmultimedia.dev (qt6Packages.qtmultimedia.dev)</li>
    <li>qt6.qtnetworkauth (qt6Packages.qtnetworkauth)</li>
    <li>qt6.qtnetworkauth.dev (qt6Packages.qtnetworkauth.dev)</li>
    <li>qt6.qtpositioning (qt6Packages.qtpositioning)</li>
    <li>qt6.qtpositioning.dev (qt6Packages.qtpositioning.dev)</li>
    <li>qt6.qtquick3d (qt6Packages.qtquick3d)</li>
    <li>qt6.qtquick3d.dev (qt6Packages.qtquick3d.dev)</li>
    <li>qt6.qtquick3dphysics (qt6Packages.qtquick3dphysics)</li>
    <li>qt6.qtquick3dphysics.dev (qt6Packages.qtquick3dphysics.dev)</li>
    <li>qt6.qtquicktimeline (qt6Packages.qtquicktimeline)</li>
    <li>qt6.qtquicktimeline.dev (qt6Packages.qtquicktimeline.dev)</li>
    <li>qt6.qtremoteobjects (qt6Packages.qtremoteobjects)</li>
    <li>qt6.qtremoteobjects.dev (qt6Packages.qtremoteobjects.dev)</li>
    <li>qt6.qtscxml (qt6Packages.qtscxml)</li>
    <li>qt6.qtscxml.dev (qt6Packages.qtscxml.dev)</li>
    <li>qt6.qtsensors (qt6Packages.qtsensors)</li>
    <li>qt6.qtsensors.dev (qt6Packages.qtsensors.dev)</li>
    <li>qt6.qtserialbus (qt6Packages.qtserialbus)</li>
    <li>qt6.qtserialbus.dev (qt6Packages.qtserialbus.dev)</li>
    <li>qt6.qtserialport (qt6Packages.qtserialport)</li>
    <li>qt6.qtserialport.dev (qt6Packages.qtserialport.dev)</li>
    <li>qt6.qtshadertools (qt6Packages.qtshadertools)</li>
    <li>qt6.qtshadertools.dev (qt6Packages.qtshadertools.dev)</li>
    <li>qt6.qtspeech (qt6Packages.qtspeech)</li>
    <li>qt6.qtspeech.dev (qt6Packages.qtspeech.dev)</li>
    <li>qt6.qtsvg (qt6Packages.qtsvg)</li>
    <li>qt6.qtsvg.dev (qt6Packages.qtsvg.dev)</li>
    <li>qt6.qttools (qt6Packages.qttools)</li>
    <li>qt6.qttools.dev (qt6Packages.qttools.dev)</li>
    <li>qt6.qttranslations (qt6Packages.qttranslations)</li>
    <li>qt6.qttranslations.dev (qt6Packages.qttranslations.dev)</li>
    <li>qt6.qtvirtualkeyboard (qt6Packages.qtvirtualkeyboard)</li>
    <li>qt6.qtvirtualkeyboard.dev (qt6Packages.qtvirtualkeyboard.dev)</li>
    <li>qt6.qtwebchannel (qt6Packages.qtwebchannel)</li>
    <li>qt6.qtwebchannel.dev (qt6Packages.qtwebchannel.dev)</li>
    <li>qt6.qtwebsockets (qt6Packages.qtwebsockets)</li>
    <li>qt6.qtwebsockets.dev (qt6Packages.qtwebsockets.dev)</li>
    <li>qt6.qtwebview (qt6Packages.qtwebview)</li>
    <li>qt6.qtwebview.dev (qt6Packages.qtwebview.dev)</li>
    <li>qt6Packages.poppler</li>
    <li>qt6Packages.poppler.dev</li>
    <li>qt6Packages.qtpbfimageplugin</li>
    <li>qt6Packages.quazip</li>
    <li>qt6Packages.quazip.dev</li>
  </ul>
</details>

**Pull request merged**